File: src/utils/atproto/tid.ts
/**
 * Wrapper around `@atproto/common-web`'s `TID` class — gives us a
 * forgiving `tidToDate()` that returns `null` for non-TID rkeys
 * (custom strings, singletons like `self`) instead of throwing.
 *
 * Using the upstream parser keeps us in lockstep with the spec; the
 * 13-char base32-sortable encoding and the (53-bit timestamp / 10-bit
 * clock-id) split are owned there.
 */

import { TID } from '@atproto/common-web';

const TID_CHAR_REGEX = /^[234567abcdefghijklmnopqrstuvwxyz]{13}$/;

/**
 * Quick syntactic check — used by callers that want to render many
 * rkeys without paying the construction cost on obviously-non-TID
 * strings. `TID.is()` upstream only checks length, not the alphabet,
 * so we pre-filter ourselves.
 */
export function looksLikeTid(input: string): boolean {
  return TID_CHAR_REGEX.test(input);
}

/**
 * Decode a TID rkey to a Date. Returns `null` when the input isn't a
 * well-formed TID or the decoded time falls outside a sane range — a
 * misencoded value could otherwise place the record in the year
 * 5000+ and the UI would happily render the nonsense.
 */
export function tidToDate(rkey: string): Date | null {
  if (!looksLikeTid(rkey)) return null;
  let micros: number;
  try {
    micros = TID.fromStr(rkey).timestamp();
  } catch {
    return null;
  }
  if (!Number.isFinite(micros) || micros <= 0) return null;
  const millis = Math.floor(micros / 1000);
  const date = new Date(millis);
  const year = date.getUTCFullYear();
  // Atproto launched in 2022. Anything well outside [2020, current+10]
  // is almost certainly a non-TID rkey that happens to pass the alphabet
  // check; treat it as not-a-TID rather than render a misleading date.
  const now = new Date().getUTCFullYear();
  if (year < 2020 || year > now + 10) return null;
  return date;
}

/**
 * Render a TID-derived date in a list-friendly way: recent dates get a
 * relative chip ("12m ago", "3d ago"); older dates get an ISO calendar
 * date so a six-month-old record doesn't read as "189d ago". Callers
 * typically pair this with a `title={date.toISOString()}` for the
 * exact timestamp on hover.
 */
export function formatTidRelative(date: Date): string {
  const diffMs = Date.now() - date.getTime();
  if (diffMs < 0) return date.toISOString().slice(0, 10);
  const sec = Math.floor(diffMs / 1000);
  if (sec < 60) return `${sec}s ago`;
  const min = Math.floor(sec / 60);
  if (min < 60) return `${min}m ago`;
  const hr = Math.floor(min / 60);
  if (hr < 24) return `${hr}h ago`;
  const day = Math.floor(hr / 24);
  if (day < 14) return `${day}d ago`;
  return date.toISOString().slice(0, 10);
}

File: ios/Packages/AturiCore/Sources/AturiCore/AtProto/TID.swift
import Foundation

/// Timestamp identifiers, the default record key. Port of
/// `src/utils/atproto/tid.ts`; the decoding the web delegates to
/// `@atproto/common-web` is done here directly.
///
/// A TID is 13 characters of base32-sortable text spelling a 64-bit value:
/// the top bit is zero, the next 53 bits are microseconds since the Unix
/// epoch, and the low 10 bits are a clock id that keeps concurrent writers
/// from colliding.
public enum TID {
    /// The base32-sortable alphabet: digits 2-7 first, then a-z, so that
    /// lexical order is chronological order.
    public static let alphabet = "234567abcdefghijklmnopqrstuvwxyz"

    private static let alphabetIndex: [Unicode.Scalar: UInt64] = {
        var table: [Unicode.Scalar: UInt64] = [:]
        for (i, scalar) in alphabet.unicodeScalars.enumerated() {
            table[scalar] = UInt64(i)
        }
        return table
    }()

    /// Quick syntactic check: 13 characters, all from the alphabet. Used by
    /// callers rendering many rkeys so obviously-non-TID strings ("self",
    /// a slug) skip the decode.
    public static func looksLikeTid(_ s: String) -> Bool {
        let scalars = s.unicodeScalars
        guard scalars.count == 13 else { return false }
        return scalars.allSatisfy { alphabetIndex[$0] != nil }
    }

    /// The 64-bit value behind a TID, or nil when it is not one. 13 base32
    /// digits carry 65 bits, so the first digit must leave the top two clear
    /// (a value below 8) for the number to be a well-formed TID at all.
    public static func rawValue(of s: String) -> UInt64? {
        guard looksLikeTid(s) else { return nil }
        var value: UInt64 = 0
        for (i, scalar) in s.unicodeScalars.enumerated() {
            let digit = alphabetIndex[scalar]!
            if i == 0, digit >= 8 { return nil }
            value = (value << 5) | digit
        }
        return value
    }

    /// Microseconds since the epoch (the top 53 bits), or nil.
    public static func microseconds(from s: String) -> UInt64? {
        rawValue(of: s).map { $0 >> 10 }
    }

    /// The low 10 bits, or nil.
    public static func clockId(from s: String) -> Int? {
        rawValue(of: s).map { Int($0 & 0x3FF) }
    }

    /// Decode a TID rkey to a Date. Nil when the input is not a well-formed
    /// TID or the decoded time falls outside a sane range: atproto launched
    /// in 2022, so anything well outside [2020, now + 10 years] is almost
    /// certainly a non-TID rkey that happens to pass the alphabet check, and
    /// rendering a date for it would mislead.
    public static func date(from rkey: String) -> Date? {
        guard let micros = microseconds(from: rkey), micros > 0 else { return nil }
        let millis = micros / 1000
        let date = Date(timeIntervalSince1970: TimeInterval(millis) / 1000)
        let year = Formatting.utcCalendar.component(.year, from: date)
        let currentYear = Formatting.utcCalendar.component(.year, from: Date())
        if year < 2020 || year > currentYear + 10 { return nil }
        return date
    }

    /// Render a TID-derived date in a list-friendly way: recent dates get a
    /// relative chip ("12m ago", "3d ago"); older dates get an ISO calendar
    /// date. See `Formatting.relative`.
    public static func formatRelative(_ date: Date, now: Date = Date()) -> String {
        Formatting.relative(date, now: now)
    }
}
